src/dns/cache.rs

```diff
@@ -0,0 +1,165 @@
+use crate::dns::{DnsCache, DnsEntry, DnsResolver, ResolveType};
+use crate::types::Result;
+use log::trace;
+use std::collections::{HashMap, VecDeque};
+
+pub(crate) struct CacheResolver {
+    values: HashMap<String, DnsEntry>,
+    max_entries: usize,
+    lru: VecDeque<String>,
+}
+
+impl DnsResolver for CacheResolver {
+    fn resolve(&mut self, domain: &str, resolve_type: ResolveType) -> Result<DnsEntry> {
+        if let Some(entry) = self.values.get(domain) {
+            if !entry.has_ipv4 && !entry.has_ipv6 && resolve_type == ResolveType::Both {
+                trace!("{}: no addresses found in entry", domain);
+                return Err(crate::types::Error::DnsNoIpAddressFound);
+            }
+            if !entry.has_ipv4 && resolve_type == ResolveType::Ipv4 {
+                trace!("{}: no ipv4 addresses found in entry", domain);
+                return Err(crate::types::Error::DnsNoIpAddressFound);
+            }
+            if !entry.has_ipv6 && resolve_type == ResolveType::Ipv6 {
+                trace!("{}: no ipv6 addresses found in entry", domain);
+                return Err(crate::types::Error::DnsNoIpAddressFound);
+            }
+
+            trace!("{}: found in cache with correct resolve type", domain);
+            self.lru.retain(|x| x != domain);
+            self.lru.push_back(domain.to_string());
+
+            return Ok(entry.clone());
+        }
+
+        Err(crate::types::Error::DnsNoIpAddressFound)
+    }
+
+    /// When a domain is resolved, it will be announced to all resolvers. This cache resolver
+    /// will store it into the cache.
+    fn announce(&mut self, domain: &str, entry: &DnsEntry) {
```

Comment:
Each resolver has an "announce", which will be called when a new dns entry has been found. It is only used by the cache resolver to store this in cache.